presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -87,4 +91,35 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                         node,
                         Assignments.identity(restrictedOutputs)));
     }
+
+    public static List<Set<Symbol>> requiredJoinInputs(JoinNode node, Set<Symbol> requiredJoinOutputs)
+    {
+        Set<Symbol> globallyUsable = requiredJoinOutputs;
+        if (node.getFilter().isPresent()) {
+            globallyUsable = ImmutableSet.<Symbol>builder()
+                    .addAll(requiredJoinOutputs)
+                    .addAll(DependencyExtractor.extractUnique(node.getFilter().get()))
+                    .build();
+        }
+
+        Set<Symbol> leftUsable = ImmutableSet.<Symbol>builder()
+                .addAll(globallyUsable)
+                .addAll(node.getCriteria().stream().map(JoinNode.EquiJoinClause::getLeft).iterator())
+                .addAll(node.getLeftHashSymbol().map(Stream::of).orElse(Stream.empty()).iterator())
+                .build();
+
+        Set<Symbol> rightUsable = ImmutableSet.<Symbol>builder()
+                .addAll(globallyUsable)
+                .addAll(node.getCriteria().stream().map(JoinNode.EquiJoinClause::getRight).iterator())
+                .addAll(node.getRightHashSymbol().map(Stream::of).orElse(Stream.empty()).iterator())
+                .build();
+
+        return ImmutableList.of(
```

Comment:
I figured out that the cross join pruning doesn't actually need to worry about this, so I can just inline the functions into the join child pruning.